modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/outgoing/OutgoingSnapshot.java

```diff
@@ -17,36 +17,269 @@
+
+    /**
+     * {@link RowId}s for which the corresponding rows were sent out of order (relative to the order in which this
+     * snapshot sends rows), hence they must be skipped when sending rows normally.
+     */
+    private final Set<RowId> overwrittenRowIds = new ConcurrentHashSet<>();
+
+    // TODO: IGNITE-17935 - manage queue size
+    /**
+     * Rows that need to be sent out of order (relative to the order in which this snapshot sends rows).
+     * Versions inside rows are in oldest-to-newest order.
+     */
+    private final Queue<SnapshotMvDataResponse.ResponseEntry> outOfOrderMvData = new LinkedList<>();
+
+    /**
+     * {@link RowId} used to point (most of the time) to the last processed row. More precisely:
+     *
+     * <ul>
+     *     <li>Before we started to read from the partition, this is equal to lowest theoretically possible
+     *     {@link RowId} for this partition</li>
+     *     <li>If we started to read from partition AND it is not yet exhausted, this is the last RowId that was
+     *     sent in this snapshot order</li>
+     *     <li>After we exhausted the partition, this is {@code null}</li>
+     * </ul>
+     */
+    private RowId lastRowId;
+
+    private boolean startedToReadPartition = false;
+
+    /**
+     * This becomes {@code true} as soon as we exhaust both the partition and out-of-order queue.
+     */
+    private boolean finished = false;
+
+    /**
+     * Creates a new instance.
+     */
+    public OutgoingSnapshot(UUID id, PartitionAccess partition, OutgoingSnapshotRegistry outgoingSnapshotRegistry) {
+        this.id = id;
+        this.partition = partition;
+        this.outgoingSnapshotRegistry = outgoingSnapshotRegistry;
+
+        lastRowId = partition.minRowId();
+    }
+
+    /**
+     * Returns the ID of this snapshot.
+     */
+    public UUID id() {
+        return id;
+    }
+
+    /**
+     * Returns the key of the corresponding partition.
+     *
+     * @return Partition key.
+     */
+    public PartitionKey partitionKey() {
+        return partition.key();
+    }
+
     /**
      * Reads a snapshot meta and returns a future with the response.
      *
      * @param metaRequest Meta request.
      */
     CompletableFuture<SnapshotMetaResponse> handleSnapshotMetaRequest(SnapshotMetaRequest metaRequest) {
-        //TODO https://issues.apache.org/jira/browse/IGNITE-17262
+        //TODO https://issues.apache.org/jira/browse/IGNITE-17935
         return null;
     }
 
     /**
      * Reads chunk of partition data and returns a future with the response.
      *
-     * @param mvDataRequest Data request.
+     * @param request Data request.
      */
-    CompletableFuture<SnapshotMvDataResponse> handleSnapshotMvDataRequest(SnapshotMvDataRequest mvDataRequest) {
-        //TODO https://issues.apache.org/jira/browse/IGNITE-17262
-        return null;
+    CompletableFuture<SnapshotMvDataResponse> handleSnapshotMvDataRequest(SnapshotMvDataRequest request) {
```

Comment:
It is a suspicious method which returns always completed future.

Reply:
I was planning to fix this in https://issues.apache.org/jira/browse/IGNITE-17935

Do you think we should do it in this PR?

Reply:
Adding a pool should be just a few lines of code, at least I think so. Why making code that you know you will change immediately after

Reply:
Added a pool